core/libp2p/crypto/hmac/hmac_provider.cpp

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "libp2p/crypto/hmac/hmac_provider.hpp"
+
+#include <openssl/conf.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+
+#include "libp2p/crypto/error.hpp"
+
+namespace libp2p::crypto::hmac {
+  using Buffer = kagome::common::Buffer;
+
+  int digestSize(common::HashType type) {
+    switch (type) {
+      case common::HashType::kSHA1:
+        return 20;
+      case common::HashType::kSHA256:
+        return 32;
+      case common::HashType::kSHA512:
+        return 64;
+      default:
+        break;
+    }
+    return 0;
+  }
+
+  const evp_md_st *makeHashTraits(common::HashType type) {
+    switch (type) {
+      case common::HashType::kSHA1:
+        return EVP_sha1();
+      case common::HashType::kSHA256:
+        return EVP_sha256();
+      case common::HashType::kSHA512:
+        return EVP_sha512();
+      default:
+        break;
+    }
+    return nullptr;
+  }
+
+  outcome::result<Buffer> HmacProvider::calculateDigest(
+      HashType hash_type, const Buffer &key, const Buffer &message) const {
+    const evp_md_st *evp_md = makeHashTraits(hash_type);
+    int digest_size = digestSize(hash_type);
+    if (evp_md == nullptr || digest_size == 0) {
+      return HmacProviderError::kUnsupportedHashMethod;
+    }
+
+    std::vector<uint8_t> result;
+    result.resize(digest_size);
+    std::fill(result.begin(), result.end(), 0);
```

Comment:
You can write the 3 lines above with one
``` std::vector<uint8_t> result(digest_size, 0); ```

Reply:
ok, done